File: bindings/java/src/main/java/io/zag/StreamingSession.java
package io.zag;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.nio.charset.StandardCharsets;
import java.util.Iterator;
import java.util.Map;
import java.util.NoSuchElementException;

/** A live streaming session with piped stdin and stdout. */
public class StreamingSession implements AutoCloseable {

    private static final ObjectMapper MAPPER = ZagProcess.MAPPER;

    private final Process process;
    private final BufferedWriter stdin;
    private final BufferedReader stdout;

    StreamingSession(Process process) {
        this.process = process;
        this.stdin =
                new BufferedWriter(
                        new OutputStreamWriter(process.getOutputStream(), StandardCharsets.UTF_8));
        this.stdout =
                new BufferedReader(
                        new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8));
    }

    /** Send a raw NDJSON line to the agent's stdin. */
    public void send(String message) throws IOException {
        stdin.write(message);
        stdin.newLine();
        stdin.flush();
    }

    /** Send a user message to the agent. */
    public void sendUserMessage(String content) throws IOException {
        String msg = MAPPER.writeValueAsString(Map.of("type", "user_message", "content", content));
        send(msg);
    }

    /** Close stdin to signal no more input. */
    public void closeInput() throws IOException {
        stdin.close();
    }

    /** Return an iterator over parsed Event objects from stdout. */
    public Iterable<Event> events() {
        return () ->
                new Iterator<>() {
                    private Event next;
                    private boolean done;

                    @Override
                    public boolean hasNext() {
                        if (done) return false;
                        if (next != null) return true;
                        try {
                            while (true) {
                                String line = stdout.readLine();
                                if (line == null) {
                                    done = true;
                                    return false;
                                }
                                String trimmed = line.trim();
                                if (trimmed.isEmpty()) continue;
                                try {
                                    next = MAPPER.readValue(trimmed, Event.class);
                                    return true;
                                } catch (Exception e) {
                                    // Skip unparseable lines
                                }
                            }
                        } catch (IOException e) {
                            done = true;
                            return false;
                        }
                    }

                    @Override
                    public Event next() {
                        if (!hasNext()) throw new NoSuchElementException();
                        Event result = next;
                        next = null;
                        return result;
                    }
                };
    }

    /** Wait for the process to exit. Throws ZagException on non-zero exit. */
    public void await() throws ZagException, InterruptedException {
        try {
            stdin.close();
        } catch (IOException ignored) {
        }

        int exitCode = process.waitFor();
        if (exitCode != 0) {
            String stderr;
            try {
                stderr =
                        new String(
                                process.getErrorStream().readAllBytes(), StandardCharsets.UTF_8);
            } catch (IOException e) {
                stderr = "";
            }
            throw new ZagException("zag exited with code " + exitCode, exitCode, stderr);
        }
    }

    @Override
    public void close() {
        process.destroyForcibly();
    }
}

File: bindings/kotlin/src/main/kotlin/zag/StreamingSession.kt
package zag

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import java.io.Closeable

/**
 * A live streaming session with piped stdin and stdout.
 */
class StreamingSession internal constructor(
    private val process: Process,
) : Closeable {

    /** Send a raw NDJSON line to the agent's stdin. */
    fun send(message: String) {
        process.outputStream.bufferedWriter().let { writer ->
            writer.write(message)
            writer.newLine()
            writer.flush()
        }
    }

    /** Send a user message to the agent. */
    fun sendUserMessage(content: String) {
        val msg = Json.encodeToString(buildJsonObject {
            put("type", "user_message")
            put("content", content)
        })
        send(msg)
    }

    /** Close stdin to signal no more input. */
    fun closeInput() {
        process.outputStream.close()
    }

    /** Async stream of parsed Event objects from stdout. */
    fun events(): Flow<Event> = flow {
        process.inputStream.bufferedReader().useLines { lines ->
            for (line in lines) {
                val trimmed = line.trim()
                if (trimmed.isEmpty()) continue

                val event = try {
                    ZagJson.decodeFromString<Event>(trimmed)
                } catch (_: Exception) {
                    continue
                }
                emit(event)
            }
        }
    }.flowOn(Dispatchers.IO)

    /** Wait for the process to exit. Throws ZagException on non-zero exit. */
    suspend fun wait(): Unit = withContext(Dispatchers.IO) {
        closeInput()

        val stderr = StringBuilder()
        val stderrThread = Thread {
            process.errorStream.bufferedReader().forEachLine { stderr.appendLine(it) }
        }.apply { isDaemon = true; start() }

        val exitCode = process.waitFor()
        stderrThread.join()

        if (exitCode != 0) {
            throw ZagException(
                "zag exited with code $exitCode",
                exitCode,
                stderr.toString(),
            )
        }
    }

    override fun close() {
        process.destroyForcibly()
    }
}
